src/test/java/testsuite/failover/FailoverIntegrationTest.java

```diff
@@ -928,140 +984,129 @@ private void assertFirstQueryThrows(Statement stmt, String expectedSQLErrorCode)
-    instanceCrasher.startCrashingInstance();
-    Thread instanceCrasherThread = new Thread(instanceCrasher);
-    instanceCrasherThread.start();
-    waitUntilInstanceIsDown(instanceId);
+  protected void makeSureInstancesUp(String... instances) throws InterruptedException {
+    makeSureInstancesUp(instances, true);
   }
 
-  /**
-   * Runnable class implementation that is used to crash an instance.
-   * */
-  public class CrashInstanceRunnable implements Runnable {
-    static final String CRASH_QUERY = "ALTER SYSTEM CRASH INSTANCE";
-    private final String instanceId;
-    private boolean keepCrashingInstance = false;
-
-    CrashInstanceRunnable(String instanceId) {
-      System.out.println("create runnable for " + instanceId);
-      this.instanceId = instanceId;
+  protected void makeSureInstancesUp(String[] instances, boolean finalCheck) throws InterruptedException {
+    this.log.logDebug("Wait until the following instances are up: \n" + String.join("\n", instances));
+    ExecutorService executorService = Executors.newFixedThreadPool(instances.length);
+    final HashSet<String> remainingInstances = new HashSet<String>();
+    remainingInstances.addAll(Arrays.asList(instances));
+
+    for (String id : instances) {
+      executorService.submit(() -> {
+        while (true) {
+          try (Connection conn = createCrashConnection(id)) {
+            conn.close();
+            remainingInstances.remove(id);
+            break;
+          } catch (SQLException ex) {
+            // Continue waiting until instance is up.
+          }
+          Thread.sleep(500);
+        }
+        return null;
+      });
     }
+    executorService.shutdown();
+    executorService.awaitTermination(120, TimeUnit.SECONDS);
 
-    public String getInstanceId() {
-      return this.instanceId;
+    if (finalCheck) {
+      assertTrue(remainingInstances.isEmpty(), "The following instances are still down: \n"
+              + String.join("\n", remainingInstances));
     }
 
-    public synchronized void stopCrashingInstance() {
-      this.keepCrashingInstance = false;
-    }
+    this.log.logDebug("The following instances are up: \n" + String.join("\n", instances));
+  }
 
-    public synchronized void startCrashingInstance() {
-      this.keepCrashingInstance = true;
-    }
+  protected void makeSureInstancesDown(String... instances) throws InterruptedException {
+    makeSureInstancesDown(instances, true);
+  }
 
-    @Override
-    public void run() {
-      while (keepCrashingInstance) {
-        try (Connection conn = createConnectionWithProxyDisabled(instanceId);
-             Statement myStmt = conn.createStatement()
-        ) {
-          myStmt.execute(CRASH_QUERY);
-        } catch (SQLException e) {
-          // Do nothing and keep creating connection to crash instance.
+  protected void makeSureInstancesDown(String[] instances, boolean finalCheck) throws InterruptedException {
```

Comment:
Looks like finalCheck is always true, so we can probably get rid of the makeSureInstancesDown methods above and remove the finalCheck boolean parameter

Reply:
reviveInstancesAndCloseTestConnection(), that executes after each test, uses finalCheck=false. Test shouldn't fail if not all instances get back online after test.

Reply:
looks like that method uses finalCheck=false for makeSureInstancesUp, I'm talking about makeSureInstancesDown here though